File: scripts/generate_data.py
#!/usr/bin/env python3
"""Parse CSV and local {genre}/{id}/ dirs -> data.json (anonymous demo)."""
from __future__ import annotations

import csv
import json
import os
import re
from pathlib import Path
from urllib.parse import parse_qs, urlparse

ROOT = Path(__file__).resolve().parent.parent
CSV_PATH = ROOT / "demo_video_links.csv"
OUT_PATH = ROOT / "data.json"

GENRES = ["documentary", "film", "lecture", "news", "vlog"]

# For certain examples, we use a local MP4 for the source video instead of YouTube.
# Keys are (genre, id) and values are filenames inside {genre}/{id}/.
# GitHub LFS rejects files > 2GB per object — keep local sources under that (compress if needed).
SOURCE_LOCAL_OVERRIDES: dict[tuple[str, str], str] = {
    ("documentary", "1"): "Mammal Origins ｜ Full Documentary ｜ NOVA ｜ PBS-23BGbVBxXdQ.mp4",
    ("documentary", "2"): "Can Dogs Talk？ ｜ Full Documentary ｜ NOVA ｜ PBS-jfLAaGtNc7U.mp4",
    ("film", "2"): "The Little Shop of Horrors 1960 Full Movie HD 1080p.mp4",
    ("vlog", "1"): "Attempting VEDA？ ｜ Meals, Planner Sticker Haul, Bathroom Organizing Project-n2YNMJShKKA.mp4",
}


def is_local_source_path(link: str) -> bool:
    t = (link or "").strip()
    return bool(t) and not re.match(r"^https?://", t, re.I) and t.endswith(".mp4")


def youtube_watch_to_embed(watch_url: str) -> str | None:
    try:
        u = urlparse(watch_url)
        q = parse_qs(u.query)
        vid = (q.get("v") or [None])[0]
        if not vid and u.netloc.replace("www.", "") in ("youtu.be",):
            vid = u.path.strip("/").split("/")[0]
        if not vid:
            return None
        return f"https://www.youtube.com/embed/{vid}"
    except Exception:
        return None


def list_mp4(d: Path) -> list[str]:
    if not d.is_dir():
        return []
    return [f.name for f in d.iterdir() if f.suffix.lower() == ".mp4"]

def pick_single(files: list[str], suffix: str) -> str | None:
    hits = [f for f in files if f.lower().endswith(suffix.lower())]
    if not hits:
        return None
    if len(hits) > 1:
        hits.sort()
    return hits[0]


def build_example(genre: str, id_str: str, youtube_embed: str | None) -> dict | None:
    dir_path = ROOT / genre / id_str
    files = list_mp4(dir_path)
    nf_files = [f for f in files if f.endswith("_no_narration.mp4")]

    keyword_clips: list[dict] = []
    for nf in nf_files:
        keyword = nf.replace("_no_narration.mp4", "").replace("_no_narration.MP4", "")
        ours_name = f"{keyword}_ours.mp4"
        if ours_name not in files:
            print(f"[skip] {dir_path}: missing {ours_name} for keyword {keyword!r}")
            continue
        keyword_clips.append({"keyword": keyword, "nf": nf, "ours": ours_name})

    keyword_clips.sort(key=lambda x: x["keyword"])
    if len(keyword_clips) < 2:
        print(f"[skip] {dir_path}: expected >=2 keywords, found {len(keyword_clips)}")
        return None

    base = f"{genre}/{id_str}"

    a2summ_name = pick_single(files, "_a2summ.mp4")
    teasergen_name = pick_single(files, "_teasergen.mp4")
    regen_name = pick_single(files, "_regen.mp4")
    fixed = {
        "a2summ": f"{base}/{a2summ_name}" if a2summ_name else None,
        "teasergen": f"{base}/{teasergen_name}" if teasergen_name else None,
        "regen": f"{base}/{regen_name}" if regen_name else None,
    }
    for k, v in fixed.items():
        if not v:
            print(f"[warn] {dir_path}: missing fixed output {k} (*{k}.mp4)")

    selected = keyword_clips[:2]
    source_local = None
    override = SOURCE_LOCAL_OVERRIDES.get((genre, id_str))
    if override:
        if override in files:
            source_local = f"{base}/{override}"
        else:
            print(f"[warn] {dir_path}: missing sourceLocal override file ({override})")
    return {
        "id": id_str,
        "youtubeEmbed": youtube_embed,
        "sourceLocal": source_local,
        "keywords": [
            {
                "keyword": s["keyword"],
                "local": {"nf": f"{base}/{s['nf']}", "ours": f"{base}/{s['ours']}"},
            }
            for s in selected
        ],
        "fixed": fixed,
    }


def main() -> None:
    rows = []
    with open(CSV_PATH, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)

    by_genre: dict[str, list] = {g: [] for g in GENRES}

    for row in rows:
        genre = (row.get("genre") or "").strip()
        id_str = str((row.get("id") or "").strip())
        link = (row.get("youtube_link") or "").strip()
        if genre not in GENRES:
            print(f"Unknown genre in CSV: {genre}")
            continue
        embed = youtube_watch_to_embed(link)
        if not embed:
            key = (genre, id_str)
            if key in SOURCE_LOCAL_OVERRIDES or is_local_source_path(link):
                embed = None  # type: ignore[assignment]
            else:
                print(f"Could not parse YouTube URL for {genre}/{id_str}")
                continue
        ex = build_example(genre, id_str, embed)
        if ex:
            by_genre[genre].append(ex)

    payload = {
        "projectTitle": "Clips2Story: Training-free Video Storyboarding and Editing using Multimodal Retrieval-Embedded Generation",
        "abstractPlaceholder": (
            "[Abstract text will appear here in the camera-ready version. "
            "This anonymous demo page is for double-blind review.]"
        ),
        "pipeline": [
            "Multimodal retrieval aligns candidate clips with story intent without task-specific training.",
            "Embedded generation composes a coherent narrative structure from retrieved visual segments.",
            "The pipeline outputs editable storyboard timelines for comparison across baselines.",
        ],
        "genres": [
            {
                "id": g,
                "label": g[0].upper() + g[1:] if g else g,
                "sets": by_genre[g],
            }
            for g in GENRES
        ],
    }

    OUT_PATH.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    print(f"Wrote {OUT_PATH}")


if __name__ == "__main__":
    main()

File: scripts/generate-data.mjs
#!/usr/bin/env node
/**
 * Parses demo_video_links.csv and local {genre}/{id}/ directories to produce data.json.
 * Anonymous demo page — no author metadata.
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const CSV_PATH = path.join(ROOT, "demo_video_links.csv");
const OUT_PATH = path.join(ROOT, "data.json");

const GENRES = ["documentary", "film", "lecture", "news", "vlog"];

// For certain examples, use a local MP4 for the source video instead of YouTube.
// Keys are `${genre}/${id}` and values are filenames inside {genre}/{id}/.
// GitHub LFS rejects files > 2GB per object — keep local sources under that (compress if needed).
const SOURCE_LOCAL_OVERRIDES = {
  "documentary/1": "Mammal Origins ｜ Full Documentary ｜ NOVA ｜ PBS-23BGbVBxXdQ.mp4",
  "film/2": "The Little Shop of Horrors 1960 Full Movie HD 1080p.mp4",
  "vlog/1": "Attempting VEDA？ ｜ Meals, Planner Sticker Haul, Bathroom Organizing Project-n2YNMJShKKA.mp4",
};

function parseCsv(text) {
  const lines = text.trim().split(/\r?\n/);
  const header = lines.shift();
  if (!header) return [];
  const rows = [];
  for (const line of lines) {
    if (!line.trim()) continue;
    // genre,id,url — URL may contain commas in rare cases; split on first two commas
    const first = line.indexOf(",");
    const second = line.indexOf(",", first + 1);
    if (first === -1 || second === -1) continue;
    const genre = line.slice(0, first).trim();
    const id = line.slice(first + 1, second).trim();
    const youtube_link = line.slice(second + 1).trim();
    rows.push({ genre, id, youtube_link });
  }
  return rows;
}

function youtubeWatchToEmbed(watchUrl) {
  try {
    const u = new URL(watchUrl);
    let videoId = u.searchParams.get("v");
    if (!videoId && (u.hostname === "youtu.be" || u.hostname === "www.youtu.be")) {
      videoId = u.pathname.replace(/^\//, "").split("/")[0];
    }
    if (!videoId) return null;
    return `https://www.youtube.com/embed/${videoId}`;
  } catch {
    return null;
  }
}

function listMp4(dir) {
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir).filter((f) => f.endsWith(".mp4"));
}

function pickSingle(files, suffix) {
  const s = suffix.toLowerCase();
  const hits = files.filter((f) => f.toLowerCase().endsWith(s));
  if (hits.length === 0) return null;
  if (hits.length > 1) hits.sort((a, b) => a.localeCompare(b));
  return hits[0];
}

function buildExampleForDir(genre, idStr, youtubeEmbed) {
  const dir = path.join(ROOT, genre, idStr);
  const files = listMp4(dir);
  const nfFiles = files.filter((f) => f.endsWith("_no_narration.mp4"));

  // Each source video now has exactly 2 target keywords. We derive them from the
  // existing NF clips in the folder (current place where keywords live).
  const keywordClips = [];
  for (const nf of nfFiles) {
    const keyword = nf.replace(/_no_narration\.mp4$/i, "");
    const oursName = `${keyword}_ours.mp4`;
    if (!files.includes(oursName)) {
      console.warn(`[skip] ${dir}: missing ${oursName} for keyword "${keyword}"`);
      continue;
    }
    keywordClips.push({ keyword, nf, oursName });
  }

  keywordClips.sort((a, b) => a.keyword.localeCompare(b.keyword));
  if (keywordClips.length < 2) {
    console.warn(`[skip] ${dir}: expected >=2 keywords, found ${keywordClips.length}`);
    return null;
  }

  const selected = keywordClips.slice(0, 2);
  const base = `${genre}/${idStr}`;

  // These 3 outputs do not change based on keyword.
  const a2summName = pickSingle(files, "_a2summ.mp4");
  const teasergenName = pickSingle(files, "_teasergen.mp4");
  const regenName = pickSingle(files, "_regen.mp4");
  const fixed = {
    a2summ: a2summName ? `${base}/${a2summName}` : null,
    teasergen: teasergenName ? `${base}/${teasergenName}` : null,
    regen: regenName ? `${base}/${regenName}` : null,
  };
  for (const [k, v] of Object.entries(fixed)) {
    if (!v) console.warn(`[warn] ${dir}: missing fixed output ${k} (*${k}.mp4)`);
  }

  let sourceLocal = null;
  const override = SOURCE_LOCAL_OVERRIDES[`${genre}/${idStr}`];
  if (override) {
    if (files.includes(override)) sourceLocal = `${base}/${override}`;
    else console.warn(`[warn] ${dir}: missing sourceLocal override file (${override})`);
  }

  return {
    id: idStr,
    youtubeEmbed,
    sourceLocal,
    keywords: selected.map((s) => ({
      keyword: s.keyword,
      local: {
        nf: `${base}/${s.nf}`,
        ours: `${base}/${s.oursName}`,
      },
    })),
    fixed: {
      a2summ: fixed.a2summ,
      teasergen: fixed.teasergen,
      regen: fixed.regen,
    },
  };
}

function main() {
  const csvRaw = fs.readFileSync(CSV_PATH, "utf8");
  const rows = parseCsv(csvRaw);

  const byGenre = {};
  for (const g of GENRES) byGenre[g] = [];

  for (const row of rows) {
    const { genre, id, youtube_link } = row;
    if (!GENRES.includes(genre)) {
      console.warn(`Unknown genre in CSV: ${genre}`);
      continue;
    }
    const embed = youtubeWatchToEmbed(youtube_link);
    if (!embed) {
      console.warn(`Could not parse YouTube URL for ${genre}/${id}`);
      continue;
    }
    const idStr = String(id);
    const ex = buildExampleForDir(genre, idStr, embed);
    if (ex) byGenre[genre].push(ex);
  }

  const payload = {
    projectTitle: "Clips2Story: Training-free Video Storyboarding and Editing using Multimodal Retrieval-Embedded Generation",
    abstractPlaceholder:
      "[Abstract text will appear here in the camera-ready version. This anonymous demo page is for double-blind review.]",
    pipeline: [
      "Multimodal retrieval aligns candidate clips with story intent without task-specific training.",
      "Embedded generation composes a coherent narrative structure from retrieved visual segments.",
      "The pipeline outputs editable storyboard timelines for comparison across baselines.",
    ],
    genres: GENRES.map((id) => ({
      id,
      label: id.charAt(0).toUpperCase() + id.slice(1),
      sets: byGenre[id] || [],
    })),
  };

  fs.mkdirSync(path.dirname(OUT_PATH), { recursive: true });
  fs.writeFileSync(OUT_PATH, JSON.stringify(payload, null, 2), "utf8");
  console.log(`Wrote ${OUT_PATH}`);
}

main();
